util/eventbus/subscribe.go

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package eventbus
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+)
+
+type dispatchFn func(vals *queue, acceptCh func() chan any) bool
+
+// A Queue receives events from a Bus.
+//
+// To receive events through the queue, see [Subscribe]. Subscribers
+// that share the same Queue receive events one at time, in the order
+// they were published.
+type Queue struct {
+	bus  *Bus
+	name string
+
+	write    chan any
+	stop     *goroutineShutdown
+	snapshot chan chan []any
+
+	outputsMu sync.Mutex
+	outputs   map[reflect.Type]dispatchFn
+}
+
+func newQueue(b *Bus, name string) *Queue {
+	ret := &Queue{
+		bus:      b,
+		name:     name,
+		write:    make(chan any),
+		stop:     newGoroutineShutdown(),
+		snapshot: make(chan chan []any),
+		outputs:  map[reflect.Type]dispatchFn{},
+	}
+	b.addQueue(ret)
+	go ret.pump()
+	return ret
+}
+
+func (q *Queue) pump() {
+	defer q.stop.ShutdownComplete()
+	var vals queue
+	acceptCh := func() chan any {
+		if vals.Full() {
+			return nil
+		}
+		return q.write
+	}
+	for {
+		if !vals.Empty() {
+			val := vals.Peek()
+			fn := q.dispatchFn(val)
+			if fn == nil {
+				// Raced with unsubscribe.
+				vals.Drop()
+				continue
+			}
+			if !fn(&vals, acceptCh) {
+				return
+			}
+		} else {
+			// Keep the cases in this select in sync with
+			// Subscriber.dispatch below. The only different should be
+			// that this select doesn't deliver queued values to
+			// anyone.
+			select {
+			case val := <-acceptCh():
+				vals.Add(val)
+			case <-q.stop.StartShutdown():
+				return
+			case ch := <-q.snapshot:
+				ch <- vals.Snapshot()
+			}
+		}
+	}
+}
+
+// A Subscriber delivers one type of event from a [Queue].
+type Subscriber[T any] struct {
+	recv *Queue
+	read chan T
+}
+
+func (s *Subscriber[T]) dispatch(vals *queue, acceptCh func() chan any) bool {
+	t := vals.Peek().(T)
+	for {
+		// Keep the cases in this select in sync with Queue.pump
+		// above. The only different should be that this select
+		// delivers a value on s.read.
+		select {
+		case s.read <- t:
+			vals.Drop()
+			return true
+		case val := <-acceptCh():
+			vals.Add(val)
+		case <-s.recv.stop.StartShutdown():
+			return false
+		case ch := <-s.recv.snapshot:
+			ch <- vals.Snapshot()
+		}
+	}
+}
+
+// Chan returns a channel on which the subscriber's events are
```

Comment:
(optional naming suggestion) Maybe "Events" since the T is arguably more important than the chan?

Reply:
Done!